test/chisel2.py

```diff
@@ -0,0 +1,229 @@
+"""
+A simple client that uses the Python ACME library to run a test issuance against
+a local Boulder server. Usage:
+
+$ virtualenv venv
+$ . venv/bin/activate
+$ pip install -r requirements.txt
+$ python chisel.py foo.com bar.com
+"""
+import json
+import logging
+import os
+import sys
+import signal
+import threading
+import time
+import urllib2
+
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.asymmetric import rsa
+
+import OpenSSL
+
+from acme import challenges
+from acme import client as acme_client
+from acme import errors as acme_errors
+from acme import jose
+from acme import messages
+from acme import standalone
+
+logging.basicConfig()
+logger = logging.getLogger()
+logger.setLevel(int(os.getenv('LOGLEVEL', 0)))
+
+DIRECTORY = os.getenv('DIRECTORY', 'http://localhost:14000/dir')
+#DIRECTORY = os.getenv('DIRECTORY', 'http://localhost:4000/directory')
+
+def make_client(email=None):
+    """Build an acme.Client and register a new account with a random key."""
+    key = jose.JWKRSA(key=rsa.generate_private_key(65537, 2048, default_backend()))
+
+    net = acme_client.ClientNetwork(key, verify_ssl=False,
+                                    user_agent="Boulder integration tester")
+
+    client = acme_client.Client(DIRECTORY, key=key, net=net)
+    tos = client.directory.meta.terms_of_service
+    if tos is not None and "Do%20what%20thou%20wilt" in tos:
+        net.account = client.register(messages.NewRegistration.from_data(email=email,
+            terms_of_service_agreed=True))
+    else:
+        raise Exception("Unrecognized terms of service URL %s" % tos)
+    return client
+
+def get_chall(authz, typ):
+    for chall_body in authz.body.challenges:
+        if isinstance(chall_body.chall, typ):
+            return chall_body
+    raise "No %s challenge found" % typ
+
+class ValidationError(Exception):
+    """An error that occurs during challenge validation."""
+    def __init__(self, domain, problem_type, detail, *args, **kwargs):
+        self.domain = domain
+        self.problem_type = problem_type
+        self.detail = detail
+
+    def __str__(self):
+        return "%s: %s: %s" % (self.domain, self.problem_type, self.detail)
+
+def make_csr(domains):
+    pkey = OpenSSL.crypto.PKey()
+    pkey.generate_key(OpenSSL.crypto.TYPE_RSA, 2048)
+    csr = OpenSSL.crypto.X509Req()
+    csr.add_extensions([
+        OpenSSL.crypto.X509Extension(
+            'subjectAltName',
+            critical=False,
+            value=', '.join('DNS:' + d for d in domains).encode()
+        ),
+    ])
+    csr.set_pubkey(pkey)
+    csr.set_version(2)
+    csr.sign(pkey, 'sha256')
+    return OpenSSL.crypto.dump_certificate_request(OpenSSL.crypto.FILETYPE_PEM, csr)
```

Comment:
We encourage folks to use the pyca/cryptography APIs for CSRs, in preference to the pyOpenSSL ones, which are pretty janky. Docs on doing what you've got here: https://cryptography.io/en/latest/x509/tutorial/#creating-a-certificate-signing-request-csr